### ISSUE TITLE
Cancellation of recovery deletes files still held onto by writes

### ISSUE THREAD

**Issue by abeyad** [elastic/elasticsearch#19473]
The RepositoryUpgradabilityIT test fails regularly because indices are deleted before trying to recreate the index for other purposes.  We could add `ensureGreen` calls after the index deletions, but the underlying cause of failure is that the ongoing recoveries get cancelled and this causes shard level data to be deleted.  However, the deletion of these files can happen *while* other threads are holding onto these files for the purposes of writes, which trips assertions where the write operations expects the temp files to be there (but they have been deleted by the recovery cancellation).

An example stack trace:
```
ERROR   21.6s J0 | RepositoryUpgradabilityIT.testRepositoryWorksWithCrossVersions <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=11223, name=elasticsearch[node_td1][generic][T#6], state=RUNNABLE, group=TGRP-RepositoryUpgradabilityIT]
   > 	at __randomizedtesting.SeedInfo.seed([6C53C46436F84D7A:76F1740152D478D]:0)
   > Caused by: java.lang.AssertionError: expected: [recovery.1468504452288._0.cfe] in []
   > 	at __randomizedtesting.SeedInfo.seed([6C53C46436F84D7A]:0)
   > 	at org.elasticsearch.indices.recovery.RecoveryTarget.assertTempFileExists(RecoveryTarget.java:407)
   > 	at org.elasticsearch.indices.recovery.RecoveryTarget.writeFileChunk(RecoveryTarget.java:397)
   > 	at org.elasticsearch.indices.recovery.RecoveryTargetService$FileChunkTransportRequestHandler.messageReceived(RecoveryTargetService.java:417)
   > 	at org.elasticsearch.indices.recovery.RecoveryTargetService$FileChunkTransportRequestHandler.messageReceived(RecoveryTargetService.java:390)
   > 	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
   > 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69)
   > 	at org.elasticsearch.transport.local.LocalTransport$1.doRun(LocalTransport.java:322)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:510)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   > 	at java.lang.Thread.run(Thread.java:745)```

**Comment by ywelsch** [comment]
Is this fixed by #19466?

**Comment by abeyad** [comment]
Yes, I just created the issue for it to link to.
